### ISSUE TITLE
Difficulty

### ISSUE THREAD

**Issue by Mojo-LB** [graft-project/GraftNetwork#91]
Dear @mbg033 ,

Since you are very active today, care to comment on: 
https://github.com/graft-project/GraftNetwork/pull/78

For a completely different approach, please check:
https://github.com/zawy12/difficulty-algorithms/issues/3 

Thank you :)

**Comment by mbg033** [comment]
@Mojo-LB we did some tests with algorithm used in Masari, but currently this task was put on hold until it's confirmed as high prio task

**Comment by Mojo-LB** [comment]
@mbg033 An argument can be made that this currently is a high priority task, looking at several issues with current network, and the implications of leaving things as they are.

However, I understand this is not your decision, and I hope the management realizes that responding to market demands is good way to approach issues.

**Comment by bobbieltd** [comment]
I object difficulty algo changes.
1. We still need to wait to see the effectiveness of new algo. 
2. Graft network is quite stable and soon big enough to avoid network fluctuation.
3. We need to wait for opinions from other projects’  well known dev.

Nothing to be hurried.

**Comment by bobbieltd** [comment]
@Mojo-LB  For info, if you use that difficulty algo, you must change Readme to put copyright notice to Massari project at Readme to avoid copyright issue. It’s really annoying when people go around to try to put their copyright at Readme of other projects.

**Comment by Mojo-LB** [comment]
@bobbieltd  Please don't make any comments before checking the linked pull request.

**Comment by bobbieltd** [comment]
@Mojo-LB Sorry, I think you are promoting zawy12’s new difficulty algo. That requires to add copyright notice to Readme and difficulty.cpp for Masari Project. One person (I guess from Masari Project) have raised copyright issue with DeroProject and I feel really angried about that demand.
For Pull Request, I have nothing to comment. Sorry again, I misunderstanded you.

**Comment by Mojo-LB** [comment]
@bobbieltd  
1. Check = check, not advertising, just what I thought is good.
2. I seen you going around spamming any post related to algos, especially zawy12. Triggered much?
3. Again, did you check the linked pull request before posting, again?

**Comment by bobbieltd** [comment]
@Mojo-LB 
1. Check = Check. Let be it.
2. I feel it’s wrong so if I see it is promoted any where. I’ll give a warning. Warning = warning, not spamming, it’s what I thought. It’s same as your way of thinking. If they didn’t raise their copyright demand, I wouldn’t need to be angried with them. I posted only two new issues at two repos where there is post about this new algo. The other two, it’s my replies like this post.
3. Before posting this, I saw your posts relating to new difficulty algo and I didn’t check your pull requests. At this point, I agree that I’m wrong. I’m not hard coders and I don’t understand much blockchain. However, I still can give out my opinion.

**Comment by Mojo-LB** [comment]
2.  I am subscribed to two repos, and I saw comment on 100% of repos I’m watching.
3. My pull request doesn’t need a coder to understand it, or to understand that is not related to Masari or zawy12. Zawy12 was an alternative suggestion.

Please have a look at it.



________________________________

**Comment by bobbieltd** [comment]
I think I’m very wrong in my over-reactions. I need to double check all infos.
For you PR, of course, I saw it and it’s unrelated.

**Comment by bobbieltd** [comment]
@Mojo-LB I think I was confused between two people which are unrelated 😂. Oufff.

**Comment by zawy12** [comment]
There are now 5 Monero clones using my LWMA algorithm.  [These charts](https://github.com/zawy12/difficulty-algorithms/issues/23)  show how much the coins have improved at stopping "hash attacks" and post-attack delays.  Miners with 25x the baseline hash rate can jump on without causing hardly any harm.  On the [LWMA page](https://github.com/zawy12/difficulty-algorithms/issues/3) you can see links to their code.

**Comment by zawy12** [comment]
The suggestion @Mojo-LB has made in his pull request to simply reduce the window from 720 to 60 is a good one.  If Monero/Cryptonote/Forknote coins had used this, they would have never been desperate to find a new algorithm, and so no one would have heard of my algorithms and kept emailing me last year to find a new one.  But the LWMA algorithm is a lot better.

**Comment by Mojo-LB** [comment]
@zawy12  Thank you for your endorsement. I realize that my pull request would be better if done as a planned fork rather than change instantly, however I made it just to show how easy it is to make things better. 

Graft Team announced they are working on it along with the monero7 asic-resistance move, but haven't announced/merged anything yet to indicate if they're simply adjusting the window or using your LWMA algo. https://www.graft.network/2018/03/22/brief-update-status-asic-resistance-difficulty-algorithm/

My personal preference is your LWMA algorithm. It has been proven repeatedly to be the best out there, and I've seen its effect on the stability of Masari network, and several others.

**Comment by zawy12** [comment]
I've been talking to ultranote, and it seems they have 2 hour delays on occasion with N=35.  So maybe it is important to go to LWMA

**Comment by Mojo-LB** [comment]
It is. 

Although adjusting difficulty window makes it much less profitable for anyone to perform a "high hashrate attack", it doesn't allow the network to efficiently recover following such incident. That is a big part where your algo really shines. 

Devs already have access to this information for some time, and are aware of the conditions that have been affecting mining since day 1 (difficulty and hash pumps) and day 30 (botnet). 

I think the real issue here is not HOW, but IF and WHEN.

**Comment by zawy12** [comment]
@bobbieltd No one has to use the Masari version of the algorithm pseudocode.  My copyright notice is not important as long as it remains open source as the M.I.T license requires.  But I like people to know where the algorithm comes from so that they can check for updated versions.   Same thing in regards to including a reference to Masari.  There are important reasons for people to know where the code originally comes from that is not related to boosting the dev's reputation or whatever.  They may want to be able to do searches that can reveal if someone else (not necessarily Masari) made an improvement or found a security hole.

**Comment by zawy12** [comment]
I hear graft lost 340 blocks in the past hour. Similar things were happening to the 20 coins that have now switched to my LWMA difficulty algorithm.   I sent what looked like the devs in discord channel an invite to our private dev discord, but they were about the only devs that didn't join.  Coins that have switched or are in process of switching:  

1. karbo
2. stellite
3. masari
4. alloy
5. dero
6. iridium
7. bitsum
8. qwerty
9. brazuk
10. ultranote
11. BTC Gold
12. BTC Candy
13. IPBC
14. turtle
15. haven
16. myztic
17. wownero
18. niobio
19. bbscoin
20. inetnse(?)
21. forknote is going to get it so this stops happening to it's offspring.

**Comment by Mojo-LB** [comment]
@zawy12, Graft haven't haven't moved to LWMA yet. Fork is planned on Block 67850 as of now (might be sooner). Changes are implemented in https://github.com/graft-project/GraftNetwork/pull/100

**Comment by zawy12** [comment]
This shows graft has been a very typical cryptonote coin, suffering from the slow difficulty at the beginning, then seeming to be OK for a while, and then slowly advancing to the point of "blowing up" from the oscillations.  They are switching just in time, although I can see the past week (last plot) was not good.   

![image](https://user-images.githubusercontent.com/18004719/38777423-7835ea0e-4075-11e8-9b7f-bd46b3179310.png)

**Comment by zawy12** [comment]
This is the chart of what happened immediately after I posted the warning above.  Graft "blew up" like the others I've been following.  This is from being too slow to respond in the presence of big miners having the option to come and go.  Maybe ASICs no longer having Monero are the overwhelming cause of this. 
 I've tried to warn Monero about this aspect of their incredibly slow algorithm, but they have no interest.  I guess they'll be alright as long as they are one of the top two coins for their POW.  But there are other reasons to have it faster, and no benefit to having it so slow.  

![image](https://user-images.githubusercontent.com/18004719/39059055-86154dc6-448b-11e8-8928-d7a2ec420e05.png)

At the worst of it (second-to-last magenta peaks) the on-off mining like this got 500 blocks in 8.3 minutes. This was followed by a 6 hour block, 20 hours to get 10 confirmations.   They then switched to LWMA and everything looks fine.    I included per-LWMA data from above because I needed to change the scales.

![image](https://user-images.githubusercontent.com/18004719/39061172-72ab6ca6-4491-11e8-87ef-2338b89b48da.png)

**Comment by zawy12** [comment]
Timestamp manipulation:  let a miner with 2% of the network hashrate send a timestamp 7000 seconds into the future once every 2 hours to see the effect.  I've told miners to stay on graft because they'll get 7 days worth of blocks in 3 hours if and when the timestamp manipulator drops by this coin like he has on 4 others the past 2 months. Terminology: time warp attack as it was originally a problem in BTC and described is not possible on coins with rolling averages.

Stable difficulty = unstable network. There is no problem from difficulty jumping all around if it that's the result of it accurately tracking network hashrate.  It also achieves more consistent solvetimes.  See the blow up above which is the result of trying to achieve a stable difficulty.   The goal is not stable difficulty, but to match difficulty to the current hashrate.  A 24 hour avg like monero is matching difficulty to the hash rate as it was 12 hours in the past (24 hr average).    Miners do not want constant difficulty.  They want a fair difficulty.  Constant difficulty = unfair mining conditions.  A difficulty that matches hashrate = a consistently fair difficulty.  Talk to miners on the 10 coins using LWMA.  They do not about it rising 3x in 30 minutes in response to a 10x hash attack , or about it dropping 80% in 30 minutes after the attackers leave.  They love it.  They are the small miners who spread the word about a coin and hold the coins, rather than a big miner looking to sell right away.   They do not have a problem with it changing all the time +/- 15% as it tries to find the correct value.

**Comment by zawy12** [comment]
I forgot that they have a limit inside the difficulty algorithm also. I made it 7 where they have a 10 in order to be more strict. The 10 is what I said only at the very beginning. That means it takes a 15% or maybe 30% minor with the LWMA to consistently lower it until the 7200 second limit is reached. I could run the test to know how low they can get it but I think someone is going to start doing this like they have been on the other coins before I get a chance to run the test..

**Comment by jagerman** [comment]
> I've told miners to stay on graft because they'll get 7 days worth of blocks in 3 hours if and when the timestamp manipulator drops by this coin like he has on 4 others the past 2 months. 

> I could run the test to know how low they can get it but I think someone is going to start doing it before I get a chance to run the test.

Indeed you appear to be several hours too late already: this has been happening to graft since last night, with the +2h forged timestamps hitting approximately every 3 hours.  The most recent is four blocks in 66220-66228; another couple at 66197 and 66204; another five forged timestamps between 66145-66154; 1 at 66109; another at 65940; two more at 65910-65911.

**Comment by imperdin** [comment]
I still firmly believe the the correct approach to solve the "Hash attacks" and the threat of 51% attacks is to slightly tweak the PoW to limit the means of acquiring massive amount of hash power

**Comment by plavirudar** [comment]
@zawy12 You're right about timewarp attacks being nomenclaturally unsound since they can also be used to refer to an older, different attack. Let's stick to using the term "forged timestamps" instead when talking about the issue currently occurring to low hashrate CN coins. 

You're also right in that miners want a fair difficulty, and when difficulty is divorced from true hashrate, there is a problem. 

The issue with using fast DAAs is that **any** sampling of a Poisson process with small N will result in fluctuations, and the only way to correct for this is to increase N. In fact, one can model this deviation. For a given coin taking N blocks into account, Poisson statistics will mean that the actual standard deviation time of these N blocks will be $sqrt(N)$, or roughly 24% from the normal after 17 blocks, or 10% from the normal after 100 blocks in the absence of true hashrate fluctuations. 

By setting N too small in the absence of external hash attacks, you end up *creating* deviations from the true hashrate, instead of protecting the network from deviations.

**Comment by jagerman** [comment]
@zawy12 - any insights into what just happened in graft, starting from 66250?  (If using the blockchain explorer website, the displayed age time looks a little wonky until you realize that it can't show negative ages--it just shows them as positives, but the age goes *down* if you refresh the page until it hits 0 after 2 hours).

**Comment by imperdin** [comment]
>Hash attacks by ASICs and NiceHase are not a problem for LWMA when it's implemented  
> with the tighter 500 second limit. >51% hash rate is only a danger to other aspects of a coin.  
  
@zawy12 We can agree that the primary goal of LWMA is to stop Hash attacks  
And one of best practices to stop attack is to chomp down the root of the problem where in our case the main offender is Nicehash.  
  
In order stop Nicehash it requires slight modifications to the PoW code and break Cryptonightv7 compatibility while "enforcing" that Nicehash will **not** support the new variant under the pretext that they are actively aiding attackers to exploit the vulnerabilities in the code.

**Comment by zawy12** [comment]
Unfortunately I have not looked at any blocks since this morning...maybe later tonight.  

Sumo and Karbo had these same problems in 2016 and early 2017, caused by the slow algorithm.  I believe that was before Nicehash was such a problem.   The POW "fix" is basically trying to minimize the number miners and their source of mining equipment.  I think it's at best a band aid, if not pouring poison on a wound.  ASICs and the same POW for everyone seems to be a legitimate market option that causes no problem for LWMA-like algorithms.

**Comment by imperdin** [comment]
> @zawy12 The POW "fix" is basically trying to minimize the number miners and their source of mining equipment.  
  
The opposite by tweaking the PoW it will further incentive legitimate miners to mine Graft as the rewards per individual miners will increase as a smaller portion of the block rewards will be "stolen" by Hash rentals.  
  
> before Nicehash was such a problem  
  
At least we can acknowledge that Nicehash is a problem.  
  
Let's take another look at the "hash stability" of Sumo / Haven.  
as far as I've seen/read they haven't had any major hash attacks after their last fork.

**Comment by jagerman** [comment]
> In fact, one can model this deviation. For a given coin taking N blocks into account, Poisson statistics will mean that the actual standard deviation time of these N blocks will be $sqrt(N)$, or roughly 24% **from the normal** after 17 blocks, or 10% from the normal after 100 blocks in the absence of true hashrate fluctuations.

One minor stats tangent on modeling this: the distribution of the time for N blocks at difficulty D precisely follows a gamma(N, 1/D)  distribution, which of course has the standard deviations you point out (that holds for any sum of independent draws), but is still noticeably different from normal at N=17:
![image](https://user-images.githubusercontent.com/4459524/39076952-db870814-44d4-11e8-82e2-b0770c54e0fe.png)
and, while closer, is still discernibly non-normal at N=100:
![image](https://user-images.githubusercontent.com/4459524/39076986-08110aa6-44d5-11e8-99c4-ba51d6e35599.png)

**Comment by imperdin** [comment]
> 51% attack was sort of theoretical  
  
Say that to [Intense](https://intensecoin.com/2018/04/16/imminent-hard-fork-and-network-attack/).  
  
Tweaking the PoW is by no means exclusive it merely imposes a slight penalty for algorithm switching and mainly Nicehash.  
Letting a coin like Graft to have 20-40% of the blocks rewards "stolen" due to hash spikes is by no means inclusive and inviting. 
  
To further emphasize the point letting the current conditions persist is like yielding to the "Viking expansion" and getting plowed for resources to further feed the malicious invaders and penalizing benevolent actors by cutting their gains. 
  
The correct approach is to create a balanced ecosystem where malicious actors are penalized instead of getting rewarded in a similar method to Minarchism which also advocates for open-market and minimal tariffs but also emphasis that "bad" actors needs to be properly penalized
  
> Minarchism is a libertarian political philosophy which advocates for the state to exist solely to provide a very small number of services. A popular model of the state proposed by minarchists is known as the night-watchman state, in which the only governmental functions are to protect citizens from aggression, theft, breach of contract and fraud as defined by property laws, limiting it to three institutions: the military, the police and courts.

**Comment by zawy12** [comment]
Again, LWMA-like DAs and tighter time limits are all that was needed to stop a significant number of "stolen" blocks. The hash attack problem is fixed.  POW work is needed for other reasons.

Hash attacks are not breaking any rule of law.  They love minarchy.  It's merely a technical challenge (that LWMA solves) to make sure they are allowed to act in their own best interest as long as they are  not lying or stealing.  They get a "steal of a deal" only when they find "stores" with a block sale going on (low difficulty).  They get the cheap stuff and future customers have to pay a higher price in part to support the coupons and discount days.  I use the "stolen" terminology in this "steal of a deal" sense.

But bad timestamps are 1) a lie 2) for profit 3) at the expense of others.  These are the 3 elements needed to prove fraud. But it seems like we have this worked out too and it's not a problem anymore.

Assuming we subconsciously don't want concentrated mining merely because we like lots of little miners in message boards promoting our coin (a big assumption), only non-difficulty 51% problems remain.

**Comment by imperdin** [comment]
> significant number of "stolen" blocks.  
  
If attackers would continue this attack cycles for all of the LWMA + SMA Cryptonightv7 coins it would amount to a massive amount of stolen blocks.  
  
P.S. I view hash attack as a rule violation, Miners are meant to secure the network not maintain it by spiking the hash-rate they are compromising the stability very chain they are mining on!.

**Comment by imperdin** [comment]
@zawy12  Please take a look at [Graft Block Explorer](https://blockexplorer.graft.network/) and tell me if it looks like a normal and healthy chain

**Comment by jagerman** [comment]
> Please take a look at Graft Block Explorer and tell me if it looks like a normal and healthy chain

Something very strange happened here.  But some reason, for All the forged timestamps somehow prevented other pools from finding any blocks for 67 blocks (66303 through 66369).  Not *one single block* in that range, over a 3 hour span, was found by any of graft's big (public) pools, despite the pools being in sync and difficulty being lower the post-fork average.  This absurd effort levels on all of them *at once* suggest something else was seriously wrong:

397.9% - https://graft.hashvault.pro/en/#!/blocks
771% - https://graftpool.net/#/blocks
9431% - https://graft.spacepools.org/#pool_blocks
1165.8% (and counting) - https://cryptoknight.cc/graft/#
589% - https://easyhash.io/graft/blocks

(The exceptionally high spacepools one appears to have had a ~30MH NiceHash order started to take advantage of the low difficulty.  Apparently it didn't work out).

It isn't at all clear why this happened, nor why it ended (with https://blockexplorer.graft.network/block/66370, found at https://graftpool.net/#/blocks).  This is more than just difficulty manipulation: difficult *wasn't* high here, yete some unknown pool made off with around 114K GRFT.

**Comment by zawy12** [comment]
Intense, like several other coins, had a problem because they copied Sumo code without knowing to lower the FTL=7200 like Sumo.  A bunch of coins got hit like someone suddenly realized the wonders of timestamp manipulation.  

Intense seems to be doing good now with LWMA.  There were two really slow blocks showing as blue spikes but those were network problems.

![image](https://user-images.githubusercontent.com/18004719/39079539-c477042e-44ea-11e8-9c1c-cf710a4c462e.png)

But like you say, Graft could end up following the same route as Niobio, Intense, Karbowanec, and probably a few others I don't know about.

**Comment by zawy12** [comment]
I was playing aorund with a new "block's stolen" metric that gave 25% above with LWMA.  In the other basis that the rest of the charts are in, it's 10%.  That's still pretty bad for LWMA, but it's what I've seen in a couple of coins after they started up.   Since that post above, I see a lot of timestamp manipulation like I've been warning.  Here one series:

![image](https://user-images.githubusercontent.com/18004719/39079192-ce974d1c-44e3-11e8-94a7-aefc21396de2.png)

He got the difficulty to drop from 1.8 to 0.9 in 18 blocks because of the 7200 limit.  The limit of 10xT in the code prevented a lot more damage.  He got 8 blocks out of 25 that I can identify  (if a large negative does not follow the large forward stamp, then he got that one too...he had already reached the 7200 limit with one big one and can't assign another one.  So he's somewhere around 33% of network power.

**Comment by zawy12** [comment]
Graft can't experience the same kind of problems the sumo clones experienced.  It handles the timestamps better.  I don't think it's going to lose 500 blocks in an hour, let along 5000.  The damage possible is limited. But you can see from the above a attacker can just keep repeating the above, getting 20 low cost blocks which results in long delays. Once that comes down, he jumps on it again, sending it possibly even lower.   He can't sit there and keep difficulty low indefinitely like he did on the Sumokoins.

**Comment by jagerman** [comment]
> But you can see from the above a attacker can just keep repeating the above, getting 20 low cost blocks which results in long delays.
Right, that seems to have been the pattern 2-3 times now, and will likely repeat.  But then there is the stretch from 66303 through 66369 which makes no sense at all.

**Comment by zawy12** [comment]
Now I see what you're saying now. The fake 7200 solvetime should have soon been followed but a -7200 negative from an honest miner. But it took 66 block for the fake to show yp...as if the attacker got 66 blocks in a row.  The real time between was not weird...170 secs per block.

![image](https://user-images.githubusercontent.com/18004719/39079871-23226b66-44f1-11e8-9434-a2af9a08fed6.png)
![image](https://user-images.githubusercontent.com/18004719/39079943-9afe2674-44f2-11e8-9bec-c67e48e00c96.png)

**Comment by zawy12** [comment]
@jagerman Do you have a reference or derivation of the gamma function?

**Comment by zawy12** [comment]
Here's what graft LWMA looks like so far.     It does not look as good as other LWMA coins, but they don't start sending a lot of bad timestamps again, it should settle down.  

![image](https://user-images.githubusercontent.com/18004719/39082675-f36d2c4c-4524-11e8-89e7-4b91830c889f.png)

**Comment by jagerman** [comment]
> Do you have a reference or derivation of the gamma function?

It's the gamma distribution (the gamma function is something quite different).  The [wikipedia article](https://en.wikipedia.org/wiki/Gamma_distribution) on it is pretty good.

The (not-so) short of it is finding a block is a binomial distribution draw with an extremely low probability of success (i.e. the inverse of the difficulty) and an extremely high number of draws (i.e. the hashrate times the number of seconds to find the block), under which conditions the binomial is for all intents and purposes a Poisson distribution.  The inter-arrival time for a Poisson (i.e. the time between blocks being found) follows a exponential distribution, in our case with rate parameter H/D (where H is the hashrate per second attempting to solve it and D is the difficulty).  The time to find N blocks is thus: x1 + x2 + x3 + .. + xN., where each xi is an independent Exp(H/D) draw.  The sum of a set of N independent and identically distributed exponentials follows a Gamma(N, H/D) distribution (this is also sometimes called an Erlang distribution when N is an integer, but is exactly the same distribution).

While the sum of independent draws from *any* distribution tends to a normal distribution in the limit (i.e. as the number of terms in the sum goes to infinity), an exponential distribution is so heavily asymmetric that it takes quite a few draws to get there, which is why even after 100 draws you can still see the skew (for something like a uniform distribution it's pretty close to normal after 5 draws and virtually indistinguishable after 10 draws).

Here's the R code that produced those density plots, if it's helpful to someone:
```R
plot(function(x) dgamma(x, 17, 1), xlim=c(5,29), ylim=c(0,0.105), xaxs="i", yaxs="i", ylab="Density", main="N=17")
plot(function(x) dnorm(x, 17, sqrt(17)), xlim=c(5,29), ylim=c(0, .105), col="red", add=T)
legend("topright", legend=c("gamma(N,1)", "normal(N, sqrt(N))"), col=c("black", "red"), lty=1)

plot(function(x) dgamma(x, 100, 1), xlim=c(70,130), ylim=c(0,0.042), xaxs="i", yaxs="i", ylab="Density", main="N=100")
plot(function(x) dnorm(x, 100, sqrt(100)), xlim=c(70,130), ylim=c(0, .042), col="red", add=T)
legend("topright", legend=c("gamma(N,1)", "normal(N, sqrt(N))"), col=c("black", "red"), lty=1)
```

**Comment by sebseb7** [comment]
Why the hell coins are still switching to this way too fast algo? It's doesn't work in the presence of attackers, when people will understand? You can play with the difficulty.

"I've tried to warn Monero about this aspect of their incredibly slow algorithm, but they have no interest"

Maybe because Monero can see how LWMA coins break one after another...

"I hear graft lost 340 blocks in the past hour. Similar things were happening to the 25 coins that have switched to my LWMA difficulty algorithm in the last 2 or 3 months."

I see a pattern here...

**Comment by Mojo-LB** [comment]
@sebseb7

Hello Sebastian Green.

This difficulty algorithm is specifically targeted against people like you, who abuse low difficulty and leave miners to suffer.

That negatively affects the network you mine, as after you stop pumping, it is left with very long block times, very far from specifications.

This DAA makes it much less profitable for you to pump TF out of it, while ensures difficulty restores back to reasonable level when you stop.

**Comment by jagerman** [comment]
> "I hear graft lost 340 blocks in the past hour. Similar things were happening to the 25 coins that have switched to my LWMA difficulty algorithm in the last 2 or 3 months."
> I see a pattern here... 

Interesting that you see a pattern.  Because the 340 blocks lost in an hour happened before the fork, i.e. when Graft was still using Monero's algorithm.

**Comment by sebseb7** [comment]
"This difficulty algorithm is specifically targeted against "

When you specifically target one "group" you forget the "other", right now graft is wide open to 51% attacks. One hole closed one opened. And now?

"That negatively affects the network you mine,"

mining affects the network, thats obvious...

"This DAA makes it much less profitable for you to pump TF out of it"

As I said: one issue closed another opened. 

"while ensures difficulty restores back to reasonable level when you stop."

It also ensures one can play with the difficutly now... I don't get the point...

**Comment by jagerman** [comment]
> right now graft is wide open to 51% attacks. One hole closed one opened.

Care to elaborate how it is *now* vulnerable to 51% attacks but wasn't before the fork?

**Comment by sebseb7** [comment]
because diff is reacting WAY to fast.

**Comment by sebseb7** [comment]
makes it very easy...

**Comment by sebseb7** [comment]
because there is a reason 

https://cdn.discordapp.com/attachments/412074460775186433/437267382491414560/unknown.png

**Comment by sebseb7** [comment]
the problem before the fork was not the difficutly algo. it was that you can buy the hashrate multiple size of the network with a few clicks. the "solution" doesn't solve anything. 

If you want to have your own blockchain, you need at least your own hashalgo... it's that simple.

**Comment by sebseb7** [comment]
trying to make the diffalgo solve things it was never intented to solve will not work.

**Comment by Mojo-LB** [comment]
That's some butthurt stuff :)

**Comment by sebseb7** [comment]
"So he's somewhere around 33% of network power."

You can order any amount of hashrate neccessary...

**Comment by zawy12** [comment]
@jagerman Can you come up with some clever math to measure current hashrate most accurately given the past 60 blocks of data? LWMA comes down fast, but it seems to have a delay in going up.  I can look see that when 5 or 6 really fast blocks come in, the difficulty needs to at least double, but LWMA N=60 at least 20 blocks.  After a lot of testing of various ideas, the best way I've found is to drop the N of the LWMA from 60 to 20 if past 5 blocks average 0.22xT, then switch back to N=60 after 15 blocks.  I need to do more testing to find a way to smoothly transition back to N=60 without losing the good results because sudden transitions are usually exploitable.  The LWMA feels very unmathematical in the sense it's not taking the Poisson distribution into account. The EMA does as shown by always having a perfect avg solvetime for all N and most hash attacks without any fiddling.  SMA needs a noticeable adjustment factor that gets larger with small N, but its greatly reduced by using harmonic mean of difficulties (average of targets) instead of the average.  I tried least squares to get the slope to estimate current hashrate but it did not seem to be very good.  The LWMA seems to work good because there is a trigger point where a lot of hash power decides to come on or off-line when difficulty changes about 25%.  So at least for dropping quickly, the LWMA seems very good for dealing with this "non-linear" drop in hashrate.  But it somehow needs to have a similar way to rise without being "asymmetrical". There are a lot of if's and's and but's when choosing an algorithm, but if you can give me a mathematically sound starting point, I can possibly make it work well with the realities of bad timestamps and on-off mining.

**Comment by sebseb7** [comment]
Why use a work-in-progress difficutly algo in a production network? 

This isn't a testnet!

" I can possibly make it work well with the realities of bad timestamps and on-off mining."

you will be one step behind always. That's not how a coin shoud be operated...

I suggest to drop this experimental difficultly algo and to focus on the root cause instead.

**Comment by zawy12** [comment]
@sebseb7  in about 3 months LWMA will have more total combined time on the 10 to 20 coins than Monero's time in existence.  I expect it to continue to work without failure on these coins which are under harsh conditions. They were mostly previously made defunct by the slowness of the Monero/Cryptonote/Bytecoin algorithm.  I suspect more eyes and testing have been applied to it than any other algorithm, partly because everyone started out having to write their own code from my pseudocode.   

You give the impression that what Mojo has said is correct.  People exposing flaws in difficulty algorithms by profit seeking has given me something interesting to work on.  If what Mojo has said is true, I only feel appreciation.  I'm not someone who had a coin or profits that was being hurt.  I like solving puzzles.  For whatever reason, difficulty became my white whale.  If people want the result, that's even better.

**Comment by sebseb7** [comment]
"under harsh conditions"

it's called the market. 


"in about 3 months"

maybe it should be researched a few years first before copy&pasteing it to 20 production coins, causing lots of trouble everywhere.


"to continue to work without failure"

so far nearly all of these coins are broken or had a desaster because of this...


" I suspect more eyes and testing have been applied to it than any other algorithm"

I can assure you thats not the case. 


" partly because everyone started out having to write their own code"

right. The coins made the mistake. They just took something without thinking first. They try to solve a problem, that cannot be solved this way, no matter how hard they try. 

"People exposing flaws in difficulty algorithms by profit seeking"

it's the russians!!!

" I like solving puzzles."

Some people invested a lot of money so you can now solve puzzels with their money. Obviously you don't feel the burden.

The problem you're trying to solve using the difficutly cannot be solved with the diffcutly algo AT ALL. Change it as long as you want, make a fork every week. The attacking script will be updated to the new situation in no time. Is it too quick, diff is a toy, is it too slow diff will have harmonic ossiciallations. At the sweet spot both "attack" types work. The real issue is the instant availability of massive amounts of cheap hashrate.

**Comment by Mojo-LB** [comment]
@zawy12 Please don't waste your time on him, he is just salty because your work is cutting his abuse potential, and because I banned from my pool for poaching miners :)

**Comment by sebseb7** [comment]
" your work is cutting his abuse potential,"

Sadly, the abuse potential grows by the week.


" and because I banned from my pool for poaching miners"

you are still welcome on my pools.

**Comment by sebseb7** [comment]
![image](https://user-images.githubusercontent.com/677956/39088473-fc9a568a-45b2-11e8-9cc0-59ceec4ac990.png)

**Comment by sebseb7** [comment]
![image](https://user-images.githubusercontent.com/677956/39088480-1fc05e8e-45b3-11e8-8184-e9f9c3a520d9.png)

**Comment by Mojo-LB** [comment]
@l1cash Thank you for that. If you read up, you realize that we are discussing the various effects of the DAA  on various people.

Unlike 4chan and reddit where you can downvote people for not liking their statements, these buttons are only available to agree/disagree with a statement, not to use for personal approvals/disapprovals.

@sebseb7  Providing actual data and references would be much helpful than spamming screenshots of unverifiable drawings.

**Comment by zawy12** [comment]
@sebseb7 Show me a coin that was hurt by my algorithm.  Show me a coin that did not did not recover from a catastrophic event because of it.   IPBC is using this algorithm since block 54881 according to my notes.  Isn't this the 3rd time someone in this thread has said "look at how well this other coin is doing" when they didn't know the coin was already using LWMA?

**Comment by sebseb7** [comment]
" Show me a coin that did not did not recover from a catastrophic event because of it."

right it will recover. but a few 600 blocks reorgs and any exchnage will delist you soon. 

"IPBC is using this algorithm since block 54881 according to my notes."

and they had a lot of trouble with it. IPBC considers this algo VERY DANGEROUS, according to my notes.

"they didn't know the coin was already using LWMA?"

LWMA is no on off switch. First: most coins implement it without simulation and testing in a hurry. And more importantly: most coins implement it for the WRONG REASONS. It cannot defend against rented hashrate 2x as big as the network and you shouldn't even try, only makes it worse...

Nothing was tested properly obviously. 

```
Age [y:d:h:m:s]: 00:000:03:55:06
```

that's pretty bad...

**Comment by zawy12** [comment]
@sebseb7 IPBC initially inherited a modified version of LWMA from alloy who had copied most of it from Masari. I immediately told Masari of the error and waited until he forked to go public.  Unbeknownst to me, Alloy copied the bad version of Masari.  That's how IPBC got it.  Later when I found out the non-LWMA code had been inserted, I contacted everyone I could find.  So the problems you thought you saw from LWMA were just leftover code from Sumo.  IPBC recently forked to implement the real LWMA.

**Comment by sebseb7** [comment]
also you should consider teaching everyone to use integer math only, because FPU math just is not portable. 

" IPBC recently forked to implement the real LWMA."

which they reverted immedately and did a meesy fork 18 blocks later.

https://github.com/ipbc-dev/ipbc/commit/dfdd74b68d5cc437c62dbaf63482beda0dd6df47

**Comment by sebseb7** [comment]
before advocating to use this algo you should maybe supply a decent reference implementation including exsessive units tests. Before this is done this algo should be avoided.

**Comment by zawy12** [comment]
If double declarations are a problem, no one is noticing.  We have a reference implementation now, wownero. You're missing the reason people rushed into it with only pseudocode: they had no choice.  Their coins were broken like the graft, karbowanc, and intense charts above and they didn't know how else to fix it.

**Comment by sebseb7** [comment]
"they had no choice"

says who? I do not share that opinion.

currently valid shares are refused by the graft daemon

![image](https://user-images.githubusercontent.com/677956/39088734-9495df64-45b7-11e8-9c13-7eb9397a439d.png)

**Comment by Mojo-LB** [comment]
@sebseb7 

> currently valid shares are refused by the graft daemon
Yes, it's happening because of the attack this DAA is trying to solve.

**Comment by zawy12** [comment]
What would you have had graft do a few days ago when their coin became defunct?

**Comment by sebseb7** [comment]
no it's because of an implementation mistake

**Comment by sebseb7** [comment]
"What would you have had graft do a few days ago when their coin became defunct?"

first: the problem was obvious for months already ("days ago")

I had only changed the hash algo not the diff algo. The old diff algo would perform nicely when on your own "private" hashalgo.

The so called "honest miner" will have no problem downloading a modified version of stackxmr to mine graft with it. 

Example https://github.com/ipbc-dev/ipbc-miner
Stable nethash of 70MH proves the point.

Nicehash will NOT open a market for a hashalgo that only one coin uses, because it makes no sense. Without nicehash all observed attacks types are gone, because they all depend on it.

**Comment by sebseb7** [comment]
maybe you should look at haven also

![image](https://user-images.githubusercontent.com/677956/39088939-f5f7f932-45bb-11e8-8d85-22494d337e2b.png)

**Comment by sebseb7** [comment]
haven forked to cryptonight-heavy which is not available at nicehash. If so it would be a small market. 

The move away from nicehash complely ended the nicehash/dumping cycle distributing the coins more evenly.

**Comment by zawy12** [comment]
I guess it's only by coincidence that when haven switched to LWMA all their problems were solved.  I'm working on getting today's charts for all the coins that have LWMA.

**Comment by sebseb7** [comment]
"I guess it's only by coincidence that when haven switched to LWMA all their problems were solved. "

true. It was their switch to cryptonight heavy that solved the problems.

**Comment by zawy12** [comment]
Masari and Iridium already had LWMA and I don't think they experienced a need to change POW.  Here's Iridium.  It's not as nice and smooth as Haven's LWMA+POW, but it shows ASICs and Nicehash are not causing it a problem.  Iridium's price has been flat, and yet it's difficulty rose 4x  in the last 5 days, as if all the Nicehash and ASICs are moving to it. 

![image](https://user-images.githubusercontent.com/18004719/39089187-d5e2b628-458e-11e8-9fe4-2a42410249bc.png)

hmmm....maybe you're not simply wanting to attack coins with bad timestamps or other things like Mojo implied.  You don't want to have to compete with Nicehash and ASICs?  That's why you don't want LWMA to be a solution?

**Comment by zawy12** [comment]
My theory that the LWMA coins (that did not upgrade POW) are surviving the influx of ASICS and Nicehash very nicely is supported by Masari doing the exact same thing as Iridium, increasing 4x in about 5 days without any price increase.

![image](https://user-images.githubusercontent.com/18004719/39089444-38c92fa0-4595-11e8-8c4a-bd28be326774.png)

**Comment by zawy12** [comment]
So what was wrong with IPBC that was supposedly caused by LWMA?  IPBC has been using LWMA over a month and F Hearns and I chat sometimes.  He mentioned a problem but didn't say how they knew there was a problem (I can't see it from the blocks) and I can't see what needed changing in the 2nd fork other than my recommendation to go to tighter limits.  

![image](https://user-images.githubusercontent.com/18004719/39090046-eabe7380-45a1-11e8-8ca5-6f97d1575821.png)

**Comment by Serena-DeroProject** [comment]
@zawy12 

May I respectfully ask that you remove your reference to Dero and LWMA in as many repositories as possible?

During the attacks on Dero both of the daemons,  C++, and the new Golang were running in parallel. However, the  Golang daemon rejected the attacked blocks which was using new algorithm.

Regards,
Serena

**Comment by zawy12** [comment]
Thoughts after looking at it for a minute: 

1) They should ask me about this kind of thing.  

2)  It says derivative of cumulative difficulty but they must have meant derivative of difficulty. Derivative of cumulative would be just difficulty itself.  You don't want to make D go lower just because it's already low.

3) If they meant derivataice of D, it's a risky positive feedback loop.  Right when D is dropping fastest they're going to make it drop even faster?  

4) These kind of attempts to make it do what you want instead of strictly adhering to math that best estimates current hashrate always has undesired side effects.  That's how BCH ended up with the EDA. Other algos have been messed up from people thinking intuitively about what they want instead of just using math to do what's required.  No matter what specific equation they use, their general approach will attract unwanted hash attacks that will exactly offset any benefit, if it's done perfectly.

5) It looks like a symmetry risk that could result in issuing blocks too fast once miners start seeking profit out of it.  They need to apply its symmetrical inverse for rising equally fast.  Interesting note: I can average only the past 2 blocks for difficulty and then apply a 1.015 and 1/1.015 limits on the rise and fall of D  and it acts almost exactly like an SMA with N = 60.

6) I'm surprised it does not fall fast enough for them.  The problem is that it does not rise fast enough. If you stop miners from wanting to visit a coin in the first place, it won't need to fall fast.

7) Arctan for this type of thing seems too contrived.  If there was a k factor next to the D' to scale it to something arctan can relate to,  I would be more optimistic that it was approximating something.

8) It's harder to code and I have not hardened it against risks, but here's one of the very very few things like this that works....actually it's the only dynamic method that appears to work good after months of searching:  When avg of past 5 solvetiems < 0.22xT or > 3xT then switch to LWMA with N/3 for N/3 blocks, then switch back to LWMA with N.  I tried various constants in SMA, Digishiled, and EMA, because I thought it would be interested to mix algorithms into one, but LWMA still won.  It goes up 3x faster and it comes down 3x faster during a hash attack. The 3xT is rarely triggered because they will leave before the initial N/3 window has finished, (the one that triggered on 0.22xT), so the 3xT never occurs because it comes down fast.   Here's a picture of the results compared to LWMA.

![image](https://user-images.githubusercontent.com/18004719/39090655-184de144-45b3-11e8-8d7d-bd5a934cc553.png)

Only dynamically switching like this based on statistical aberration can help without also causing an increase in random variation that invites hash attacks.

**Comment by jagerman** [comment]
> 5. I'm surprised they say it does not fall fast enough. The bigger problem I see is that it does not rise fast enough.
I think falling not fast enough is referring to wallclock seconds, rather than number of blocks (since under a falling hashrate average block times are further apart).  I'm not sure what that does to the incentive to try to game the system, though.

**Comment by sebseb7** [comment]
"they accidentally reverted to a previous algorithm"

no with full intent. You spoke to them even once?

**Comment by sebseb7** [comment]
" They switched POW during this time, but you can't even tell when"

I can. IPBCs problems ended the second the PoW was switched. Same for Haven. There has been ZERO problems with those coins who left the monero PoW.

**Comment by Serena-DeroProject** [comment]
@jagerman 

The references to Dero have been removed. With respect, I didn't come here to publicly discuss difficulty algorithms, I came to ask that reference's to Dero please be removed, and they have been.

Dero makes no claim to the LWMA algorithm.

**Comment by wowario** [comment]
Dero, Dero, Dero, Dero,Dero, Dero, Dero, Dero, Dero, Dero!

**Comment by zawy12** [comment]
@sebseb7 I chat with IPBC's F.Hearns.  I didn't keep my word to review his paper this past week.   He said they thought an unsigned integer times a signed one was causing a problem, so they switched back from a tweaked LWMA to the previous LWMA.  He didn't go into any details. What problems were they having?  I don't see a problem in the difficulties or solvetimes.   I sent him a DM to try to find out the details.

**Comment by zawy12** [comment]
@Mojo-LB Graft did the FTL and MTP adjustments like I was wanting, so you can close this issue.

**Comment by Mojo-LB** [comment]
Thank you @zawy12  for all the help you provided and all the time you spent. 

Thank you to all the contributors who helped :)

**Comment by sebseb7** [comment]
"What problems were they having?"

the new implementation caused the deamon to not find any blocks because of a possible overun so that part of the fork was cancelled.

The plan was to fork to a new PoW and a new LWMA implementation. Only the PoW fork happened and it was successful. No more hashrate fluctuations at all.

**Comment by zawy12** [comment]
OK, I see the difficulty drop at the end was from implementation of the POW.  But the hashrate fluctuations were not causing any problem.   The blue peaks did not happen very often, and they are not tall.  That means they did not have any delays.  The perfect average solvetime shows things were functioning fine.  

I'll work on finding out what happened.  Havendev is looking into it because your PM got him worried. (they copied his code)

**Comment by sebseb7** [comment]
"But the hashrate fluctuations were not causing any problem."

Not at IPBC thats true. IPBC will work with ANY diff algo...

**Comment by sebseb7** [comment]
" Havendev is looking into it because your PM got him worried. "

The problem at IPBC should worry him. He said "I also spoke to the algo creator Zawy who double reassured there was no issue."

You double reassured when "I'll work on finding out what happened."

WIP in production how I like that.

**Comment by zawy12** [comment]
IPBC problem: Since other coins are not having a problem with it, and it worked fine in their testing, and since no pool switched to the new difficulty at the fork, and no block was found, I think it's speculative to believe the fork was done correctly and to assume there was something wrong with the algorithm.

**Comment by sebseb7** [comment]
"Why do you view hash rate fluctuations as a problem if the solvetimes are OK and the big miners were unable to get cheap blocks?"

never said hashrate fluctuations are a problem. 

big miners don't need cheap blocks, they are more effective so then can work with way lower margins.

**Comment by sebseb7** [comment]
"since no pool switched to the new difficulty at the fork, and no block was found"

- daemon and pool software was updated at the pools. 
- No block was found because of daemon bug. 
- LWMA update removed -> all fine

**Comment by sebseb7** [comment]
" I think it's speculative to believe the fork was done correctly and to assume there was something wrong with the algorithm."

Why is it speculative? You can get the code from github and replay the chain.

**Comment by sebseb7** [comment]
" and to assume there was something wrong with the algorithm"

I assume you have never seen that before:

2018-Apr-16 11:16:51.701863 ERROR   difficulty overhead.
2018-Apr-16 11:16:51.701927 ERROR   Failed to create block template

**Comment by sebseb7** [comment]
graft right now also fails to accept valid shares:

```
2018-04-22 13:56:25.414    [RPC0]  ERROR   cn      src/cryptonote_core/cryptonote_core.cpp:1045    mined block failed verification
```

**Comment by zawy12** [comment]
You implied hash rate fluctuations are bad with 

>Only the PoW fork happened and it was successful. No more hashrate fluctuations at all.

So, then.....why are you saying POW change was needed by IPBC?  What are you saying is improved by changing POW if a coin already has LWMA?  

> daemon and pool software was updated at the pools.

Why do you say that when all the miners were saying no pool would update to the newer LWMA?  It just sounded like teh fork was not done correctly.

> Why is it speculative? You can get the code from github and replay the chain.

That doesn't determine if the fork method or newer LWMA was the problem.  Obviously the newer LWMA was fine because testnet was fine.  Obviously the fork was bad because ... well it never forked to run the algorithm like testnet did.

**Comment by sebseb7** [comment]
"You implied hash rate fluctuations are bad with"
"So, then.....why are you saying POW change was needed by IPBC"

I said the PoW change made the hashrate stable. Not stated my opinion wether this is good or bad. 

"Why do you say that when all the miners were saying no pool would update to the newer LWMA?"

maybe you should ask the pools not the miners. They miners can't know the daemon version the pool is running. 

" Obviously the newer LWMA was fine because testnet was fine."

when you use a testnet with higher difficutly the testnet fails also.

" Obviously the fork was bad because ... well it never forked to run the algorithm like testnet did."

obviously you are assuming all of this.

**Comment by zawy12** [comment]
>I said the PoW change made the hashrate stable. Not stated my opinion wether this is good or bad.

Again, why did you come here saying a POW change was needed if unstable hashrate is neither good or bad?

>when you use a testnet with higher difficutly the testnet fails also.

How are you going to force testnet to a higher difficulty when it's using the correct algorithm?

**Comment by sebseb7** [comment]
"Again, why did you come here saying a POW change was needed if unstable hashrate is neither good or bad?"

those people who want to stabilize graft nethash should consider a PoW change. I don't care wether hashrate is stable or not. 

"How are you going to force testnet to a higher difficulty when it's using the correct algorithm?"

you can manupilate the speed of the clock for example... do I really need to explain this?

**Comment by zawy12** [comment]
Your presumption is that LWMA gave an unjustifiably (wrong) high difficulty due to a code error, and you're going to accelerate the testnet clock in order to give a correct high difficulty?    What does that achieve?   So you claim to have all the data to prove what you're saying.  What were the timestamps sent to the LWMA  algorithm to make it throw an unusually high difficulty?  What was the value of that high difficulty?   What difficulty did the pools think was the correct one, and based on what timestamps.  We need some sort of data to back up your speculations.

**Comment by plavirudar** [comment]
>big miners don't need cheap blocks
Big miners need cheap blocks as much as (or more so than) the average miner, since they're doing it for profit instead of doing it for a hobby.

**Comment by sebseb7** [comment]
"Your presumption is that LWMA gave an unjustifiably (wrong) high difficulty due to a code error"

no I said the IPBC daemon failed at fork time, then LWMA was removed. Thats all I said. Also I don't need to prove it because everyone interested can look for themself. The code is available, the blockchain also.

" We need some sort of data to back up your speculations."

You said you love Puzzles. Let's have some fun...

**Comment by zawy12** [comment]
>no I said the IPBC daemon failed at fork time

It seems you've reversed course away from promoting a POW change as a cure-all, and now it seems you're reluctant to directly point the finger at the LWMA code as the cause of IPBC's unique fork problem.

**Comment by sebseb7** [comment]
I don't know what went wrong at IPBC and when problems gone now. But fact is, graft may never reach fork height, because it is broken:

![image](https://user-images.githubusercontent.com/677956/39098068-87d91f4e-4665-11e8-86ab-352c733720ca.png)

**Comment by sebseb7** [comment]
:laughing: 

![image](https://user-images.githubusercontent.com/677956/39098089-c68523e6-4665-11e8-8a5a-253017ee9a05.png)

**Comment by sebseb7** [comment]
"It seems you've reversed course away from promoting a POW change as a cure-all"

I still promote it as a cure to the current situation.

**Comment by zawy12** [comment]
I [post this](https://github.com/graft-project/GraftNetwork/issues/115#issuecomment-383845401) to wrong issue.
